src/nrnpython/nrnpy_nrn.cpp

```diff
@@ -2928,23 +2922,9 @@ static PyMethodDef NPySegObj_methods[] = {
     {"volume", (PyCFunction) seg_volume_safe, METH_NOARGS, "Segment volume (um3)"},
     {NULL}};
 
-// I'm guessing Python should change their typedef to get rid of the
-// four "deprecated conversion from string constant to 'char*'" warnings.
-// Could avoid by casting each to (char*) but probably better to keep the
-// warnings. For now we get rid of the warnings by copying the string to
-// char array.
-static char* cpstr(const char* s) {
-    char* s2 = new char[strlen(s) + 1];
-    strcpy(s2, s);
-    return s2;
-}
 static PyMemberDef NPySegObj_members[] = {
-    {cpstr("x"),
-     T_DOUBLE,
-     offsetof(NPySegObj, x_),
-     0,
-     cpstr("location in the section (segment containing x)")},
-    {cpstr("sec"), T_OBJECT_EX, offsetof(NPySegObj, pysec_), 0, cpstr("Section")},
+    {"x", T_DOUBLE, offsetof(NPySegObj, x_), 0, "location in the section (segment containing x)"},
+    {"sec", T_OBJECT_EX, offsetof(NPySegObj, pysec_), 0, "Section"},
```

Comment:
Is the comment about warnings still real?

Reply:
Not w/ clang18 or g++-13; The latter now gives (prepare yourself, this is with `-Wall`):
```
In file included from /home/gevaert/tmp/nrn/src/nrnpython/../nrnoc/section.h:29,
                 from /home/gevaert/tmp/nrn/src/nrnpython/../nrniv/nrnoc2iv.h:6,
                 from /home/gevaert/tmp/nrn/src/nrnpython/nrnpy_nrn.cpp:10:
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/model_data.hpp: In member function ‘void neuron::Model::delete_mechanism(int)’:
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/model_data.hpp:80:18: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<std::unique_ptr<neuron::container::Mechanism::storage> >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
   80 |         if (type >= m_mech_data.size() || !m_mech_data[type]) {
      |             ~~~~~^~~~~~~~~~~~~~~~~~~~~
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/model_data.hpp: In member function ‘bool neuron::Model::is_valid_mechanism(int) const’:
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/model_data.hpp:108:34: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
  108 |         return 0 <= type && type < mechanism_storage_size() && m_mech_data[type];
      |                             ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/model_data.hpp: In member function ‘neuron::container::Mechanism::storage& neuron::Model::mechanism_data_impl(int) const’:
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/model_data.hpp:127:30: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<std::unique_ptr<neuron::container::Mechanism::storage> >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
  127 |         if (type < 0 || type >= m_mech_data.size()) {
      |                         ~~~~~^~~~~~~~~~~~~~~~~~~~~
/home/gevaert/tmp/nrn/src/nrnpython/nrnpy_nrn.cpp: In function ‘PyObject* var_of_mech_next(NPyVarOfMechIter*)’:
/home/gevaert/tmp/nrn/src/nrnpython/nrnpy_nrn.cpp:2166:18: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
 2166 |     if (self->i_ >= self->msym_->s_varn) {
      |         ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/home/gevaert/tmp/nrn/src/nrnpython/nrnpy_nrn.cpp: In function ‘void nrnpy_reg_mech(int)’:
/home/gevaert/tmp/nrn/src/nrnpython/nrnpy_nrn.cpp:3148:19: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
 3148 |     for (i = 0; i < mf.sym->s_varn; ++i) {
      |                 ~~^~~~~~~~~~~~~~~~
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/model_data.hpp: In instantiation of ‘void neuron::Model::apply_to_mechanisms(const Callable&) [with Callable = neuron::Model::shrink_to_fit()::<lambda(auto:26&)>]’:
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/model_data.hpp:122:28:   required from here
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/model_data.hpp:38:34: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<std::unique_ptr<neuron::container::Mechanism::storage> >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
   38 |         for (auto type = 0; type < m_mech_data.size(); ++type) {
      |                             ~~~~~^~~~~~~~~~~~~~~~~~~~
In file included from /home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/mechanism_data.hpp:3,
                 from /home/gevaert/tmp/nrn/src/nrnpython/../nrnoc/section.h:27:
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp: In instantiation of ‘Callable neuron::container::detail::field_data<Tag, neuron::container::detail::FieldImplementation::RuntimeVariable>::for_each_vector(Callable) const [with Callable = neuron::container::soa<neuron::container::Mechanism::storage, neuron::container::Mechanism::field::FloatingPoint>::size() const::<lambda(const auto:4&, const auto:5&, int, int)>; Tag = neuron::container::Mechanism::field::FloatingPoint]’:
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:885:92:   required from ‘Callable neuron::container::soa< <template-parameter-1-1>, <template-parameter-1-2> >::for_each_tag_vector_impl(Callable) const [with Callable = neuron::container::soa<neuron::container::Mechanism::storage, neuron::container::Mechanism::field::FloatingPoint>::size() const::<lambda(const auto:4&, const auto:5&, int, int)>; Tag = neuron::container::Mechanism::field::FloatingPoint; RemainingTags = {}; Storage = neuron::container::Mechanism::storage; Tags = {neuron::container::Mechanism::field::FloatingPoint}]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:880:59:   required from ‘Callable neuron::container::soa< <template-parameter-1-1>, <template-parameter-1-2> >::for_each_vector(Callable) const [with Callable = neuron::container::soa<neuron::container::Mechanism::storage, neuron::container::Mechanism::field::FloatingPoint>::size() const::<lambda(const auto:4&, const auto:5&, int, int)>; Storage = neuron::container::Mechanism::storage; Tags = {neuron::container::Mechanism::field::FloatingPoint}]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:742:24:   required from ‘std::size_t neuron::container::soa< <template-parameter-1-1>, <template-parameter-1-2> >::size() const [with Storage = neuron::container::Mechanism::storage; Tags = {neuron::container::Mechanism::field::FloatingPoint}; std::size_t = long unsigned int]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/model_data.hpp:83:54:   required from here
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:483:28: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<std::vector<double>, std::allocator<std::vector<double> > >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
  483 |         for (auto i = 0; i < m_storage.size(); ++i) {
      |                          ~~^~~~~~~~~~~~~~~~~~
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp: In instantiation of ‘Callable neuron::container::detail::field_data<Tag, neuron::container::detail::FieldImplementation::RuntimeVariable>::for_each_vector(Callable) [with neuron::container::detail::may_cause_reallocation might_reallocate = neuron::container::detail::may_cause_reallocation::Yes; Callable = neuron::container::soa<neuron::container::Mechanism::storage, neuron::container::Mechanism::field::FloatingPoint>::shrink_to_fit()::<lambda(const auto:8&, auto:9&, int, int)>; Tag = neuron::container::Mechanism::field::FloatingPoint]’:
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:853:90:   required from ‘Callable neuron::container::soa< <template-parameter-1-1>, <template-parameter-1-2> >::for_each_tag_vector_impl(Callable) [with neuron::container::detail::may_cause_reallocation might_reallocate = neuron::container::detail::may_cause_reallocation::Yes; Callable = neuron::container::soa<neuron::container::Mechanism::storage, neuron::container::Mechanism::field::FloatingPoint>::shrink_to_fit()::<lambda(const auto:8&, auto:9&, int, int)>; Tag = neuron::container::Mechanism::field::FloatingPoint; RemainingTags = {}; Storage = neuron::container::Mechanism::storage; Tags = {neuron::container::Mechanism::field::FloatingPoint}]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:844:77:   required from ‘Callable neuron::container::soa< <template-parameter-1-1>, <template-parameter-1-2> >::for_each_vector(Callable) [with neuron::container::detail::may_cause_reallocation might_reallocate = neuron::container::detail::may_cause_reallocation::Yes; Callable = neuron::container::soa<neuron::container::Mechanism::storage, neuron::container::Mechanism::field::FloatingPoint>::shrink_to_fit()::<lambda(const auto:8&, auto:9&, int, int)>; Storage = neuron::container::Mechanism::storage; Tags = {neuron::container::Mechanism::field::FloatingPoint}]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:771:61:   required from ‘void neuron::container::soa< <template-parameter-1-1>, <template-parameter-1-2> >::shrink_to_fit() [with Storage = neuron::container::Mechanism::storage; Tags = {neuron::container::Mechanism::field::FloatingPoint}]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/model_data.hpp:122:74:   required from ‘neuron::Model::shrink_to_fit()::<lambda(auto:26&)> [with auto:26 = neuron::container::Mechanism::storage]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/model_data.hpp:42:21:   required from ‘void neuron::Model::apply_to_mechanisms(const Callable&) [with Callable = neuron::Model::shrink_to_fit()::<lambda(auto:26&)>]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/model_data.hpp:122:28:   required from here
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:466:28: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<std::vector<double>, std::allocator<std::vector<double> > >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
  466 |         for (auto i = 0; i < m_storage.size(); ++i) {
      |                          ~~^~~~~~~~~~~~~~~~~~
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp: In instantiation of ‘Callable neuron::container::detail::field_data<Tag, neuron::container::detail::FieldImplementation::RuntimeVariable>::for_each_vector(Callable) [with neuron::container::detail::may_cause_reallocation might_reallocate = neuron::container::detail::may_cause_reallocation::Yes; Callable = neuron::container::soa<neuron::container::Mechanism::storage, neuron::container::Mechanism::field::FloatingPoint>::acquire_owning_identifier()::<lambda(const auto:18&, auto:19&, auto:20, auto:21)>; Tag = neuron::container::Mechanism::field::FloatingPoint]’:
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:853:90:   required from ‘Callable neuron::container::soa< <template-parameter-1-1>, <template-parameter-1-2> >::for_each_tag_vector_impl(Callable) [with neuron::container::detail::may_cause_reallocation might_reallocate = neuron::container::detail::may_cause_reallocation::Yes; Callable = neuron::container::soa<neuron::container::Mechanism::storage, neuron::container::Mechanism::field::FloatingPoint>::acquire_owning_identifier()::<lambda(const auto:18&, auto:19&, auto:20, auto:21)>; Tag = neuron::container::Mechanism::field::FloatingPoint; RemainingTags = {}; Storage = neuron::container::Mechanism::storage; Tags = {neuron::container::Mechanism::field::FloatingPoint}]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:844:77:   required from ‘Callable neuron::container::soa< <template-parameter-1-1>, <template-parameter-1-2> >::for_each_vector(Callable) [with neuron::container::detail::may_cause_reallocation might_reallocate = neuron::container::detail::may_cause_reallocation::Yes; Callable = neuron::container::soa<neuron::container::Mechanism::storage, neuron::container::Mechanism::field::FloatingPoint>::acquire_owning_identifier()::<lambda(const auto:18&, auto:19&, auto:20, auto:21)>; Storage = neuron::container::Mechanism::storage; Tags = {neuron::container::Mechanism::field::FloatingPoint}]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:1171:61:   required from ‘neuron::container::owning_identifier<Storage> neuron::container::soa< <template-parameter-1-1>, <template-parameter-1-2> >::acquire_owning_identifier() [with Storage = neuron::container::Mechanism::storage; Tags = {neuron::container::Mechanism::field::FloatingPoint}]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_identifier.hpp:73:62:   required from ‘neuron::container::owning_identifier< <template-parameter-1-1> >::owning_identifier(Storage&) [with Storage = neuron::container::Mechanism::storage]’
/home/gevaert/tmp/nrn/src/nrnpython/../nrnoc/section.h:236:53:   required from here
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:466:28: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<std::vector<double>, std::allocator<std::vector<double> > >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp: In instantiation of ‘Callable neuron::container::detail::field_data<Tag, neuron::container::detail::FieldImplementation::RuntimeVariable>::for_each_vector(Callable) [with neuron::container::detail::may_cause_reallocation might_reallocate = neuron::container::detail::may_cause_reallocation::No; Callable = neuron::container::soa<neuron::container::Mechanism::storage, neuron::container::Mechanism::field::FloatingPoint>::erase(std::size_t)::<lambda(const auto:10&, auto:11&, int, int)>; Tag = neuron::container::Mechanism::field::FloatingPoint]’:
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:853:90:   required from ‘Callable neuron::container::soa< <template-parameter-1-1>, <template-parameter-1-2> >::for_each_tag_vector_impl(Callable) [with neuron::container::detail::may_cause_reallocation might_reallocate = neuron::container::detail::may_cause_reallocation::No; Callable = neuron::container::soa<neuron::container::Mechanism::storage, neuron::container::Mechanism::field::FloatingPoint>::erase(std::size_t)::<lambda(const auto:10&, auto:11&, int, int)>; Tag = neuron::container::Mechanism::field::FloatingPoint; RemainingTags = {}; Storage = neuron::container::Mechanism::storage; Tags = {neuron::container::Mechanism::field::FloatingPoint}]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:844:77:   required from ‘Callable neuron::container::soa< <template-parameter-1-1>, <template-parameter-1-2> >::for_each_vector(Callable) [with neuron::container::detail::may_cause_reallocation might_reallocate = neuron::container::detail::may_cause_reallocation::No; Callable = neuron::container::soa<neuron::container::Mechanism::storage, neuron::container::Mechanism::field::FloatingPoint>::erase(std::size_t)::<lambda(const auto:10&, auto:11&, int, int)>; Storage = neuron::container::Mechanism::storage; Tags = {neuron::container::Mechanism::field::FloatingPoint}]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:798:64:   required from ‘void neuron::container::soa< <template-parameter-1-1>, <template-parameter-1-2> >::erase(std::size_t) [with Storage = neuron::container::Mechanism::storage; Tags = {neuron::container::Mechanism::field::FloatingPoint}; std::size_t = long unsigned int]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_identifier.hpp:159:33:   required from ‘void neuron::container::owning_identifier< <template-parameter-1-1> >::destroy() [with Storage = neuron::container::Mechanism::storage]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_identifier.hpp:95:9:   required from ‘neuron::container::owning_identifier< <template-parameter-1-1> >::~owning_identifier() [with Storage = neuron::container::Mechanism::storage]’
/home/gevaert/tmp/nrn/src/nrnpython/../nrnoc/section.h:236:53:   required from here
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:466:28: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<std::vector<double>, std::allocator<std::vector<double> > >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp: In instantiation of ‘Callable neuron::container::detail::field_data<Tag, neuron::container::detail::FieldImplementation::RuntimeVariable>::for_each_vector(Callable) [with neuron::container::detail::may_cause_reallocation might_reallocate = neuron::container::detail::may_cause_reallocation::No; Callable = neuron::container::soa<neuron::container::Mechanism::storage, neuron::container::Mechanism::field::FloatingPoint>::erase(std::size_t)::<lambda(const auto:12&, auto:13&, int, int)>; Tag = neuron::container::Mechanism::field::FloatingPoint]’:
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:853:90:   required from ‘Callable neuron::container::soa< <template-parameter-1-1>, <template-parameter-1-2> >::for_each_tag_vector_impl(Callable) [with neuron::container::detail::may_cause_reallocation might_reallocate = neuron::container::detail::may_cause_reallocation::No; Callable = neuron::container::soa<neuron::container::Mechanism::storage, neuron::container::Mechanism::field::FloatingPoint>::erase(std::size_t)::<lambda(const auto:12&, auto:13&, int, int)>; Tag = neuron::container::Mechanism::field::FloatingPoint; RemainingTags = {}; Storage = neuron::container::Mechanism::storage; Tags = {neuron::container::Mechanism::field::FloatingPoint}]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:844:77:   required from ‘Callable neuron::container::soa< <template-parameter-1-1>, <template-parameter-1-2> >::for_each_vector(Callable) [with neuron::container::detail::may_cause_reallocation might_reallocate = neuron::container::detail::may_cause_reallocation::No; Callable = neuron::container::soa<neuron::container::Mechanism::storage, neuron::container::Mechanism::field::FloatingPoint>::erase(std::size_t)::<lambda(const auto:12&, auto:13&, int, int)>; Storage = neuron::container::Mechanism::storage; Tags = {neuron::container::Mechanism::field::FloatingPoint}]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:807:60:   required from ‘void neuron::container::soa< <template-parameter-1-1>, <template-parameter-1-2> >::erase(std::size_t) [with Storage = neuron::container::Mechanism::storage; Tags = {neuron::container::Mechanism::field::FloatingPoint}; std::size_t = long unsigned int]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_identifier.hpp:159:33:   required from ‘void neuron::container::owning_identifier< <template-parameter-1-1> >::destroy() [with Storage = neuron::container::Mechanism::storage]’
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_identifier.hpp:95:9:   required from ‘neuron::container::owning_identifier< <template-parameter-1-1> >::~owning_identifier() [with Storage = neuron::container::Mechanism::storage]’
/home/gevaert/tmp/nrn/src/nrnpython/../nrnoc/section.h:236:53:   required from here
/home/gevaert/tmp/nrn/src/nrnpython/../neuron/container/soa_container.hpp:466:28: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<std::vector<double>, std::allocator<std::vector<double> > >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
/home/gevaert/tmp/nrn/src/nrnpython/nrnpy_nrn.cpp: At global scope:
/home/gevaert/tmp/nrn/src/nrnpython/nrnpy_nrn.cpp:3154:6: warning: ‘void nrnpy_unreg_mech(int)’ defined but not used [-Wunused-function]
 3154 | void nrnpy_unreg_mech(int type) {
      |      ^~~~~~~~~~~~~~~~
/home/gevaert/tmp/nrn/src/nrnpython/nrnpy_nrn.cpp:251:13: warning: ‘void NPySecObj_dealloc_safe(NPySecObj*)’ defined but not used [-Wunused-function]
  251 | static void NPySecObj_dealloc_safe(NPySecObj* self) {
      |             ^~~~~~~~~~~~~~~~~~~~~~
```

Reply:
Okay, then it seems like an easy simplification. Thank you.

Reply:
Here's the commit that introduces the construct that's being removed:
https://github.com/neuronsimulator/nrn/commit/e4e50b578a9169f327c20fbef32629ff511007b1